### PR TITLE
feat: enhance binary array encoding, make it the default

### PR DESCRIPTION
**Pull request by westonpace**
This adds support for the following things to the binary encoding:
 * Nulls
 * Large offsets
 * Different types (e.g. String vs Binary)

In addition, I have changed the row limit on pages from `u32` to `u64`.  Partly, this is because string arrays larger that 2GB are not unheard of and it may be nice in some cases to store them in one page.  Also pages are infrequent enough that an extra 4 bytes is trivial (actually, protobuf already stores u32 and u64 the same).  Another reason is that, with some encodings (e.g. constant / compressed bitmap / etc.) it's possible to have 2^64 rows without occupying a large amount of space.

Note: this is technically a `.proto` change but in protobuf `u32 -> u64` is backwards compatible so it is not a huge issue.

### CODE REVIEW

File: rust/lance-encoding/src/encodings/physical/binary.rs
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/binary.rs:249]:
```suggestion
        // Buffers[2] == null buffer // TODO: Micro-optimization, can we get rid of this?  Doesn't hurt much though
        //                              This buffer is always empty
        // Buffers[3] == bytes buffer
```

File: rust/lance-encoding/src/encodings/physical/binary.rs
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/binary.rs:253]:
```suggestion
```

File: rust/lance-encoding/src/encodings/physical/binary.rs
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/binary.rs:512]:
```suggestion
    async fn test_utf8() {
        let field = Field::new("", DataType::Utf8, false);
        check_round_trip_encoding_random(field).await;
    }
```

**Comment by raunaks13** [comment]
I think we can merge this after #2508 is merged in

**Comment by raunaks13** [comment]
I think we can merge this after #2508 is merged in

**Comment by raunaks13** [comment]
I think we can merge this as soon as #2508 is merged in

**Review by raunaks13 — APPROVED** [review]
I think we can merge this as soon as #2508 is merged in

**Pull request merged**
